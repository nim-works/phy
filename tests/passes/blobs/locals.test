discard """
  description: "
    Ensure that locals of non-blob type that have their address taken are
    turned into blob locals and that parameters are handled correctly
  "
  output: "(Done)"
"""
(TypeDefs
  (ProcTy (Void) (Int 4))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 0)
    (Locals (Int 4))
    (List
      (Block (Params (Local 0))
        (Asgn (Local 0) (IntVal 0))
        (Drop (Copy (Local 0)))
        (Drop (Addr (Local 0)))
        (Return))))

  (ProcDef (Type 1)
    (Locals (Int 8))
    (List
      (Block (Params)
        (Drop (Addr (Local 0)))
        (Return))
      (Except (Params (Local 0))
        (Return)))))
