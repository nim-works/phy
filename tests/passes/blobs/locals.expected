;$sexp
(TypeDefs
  (ProcTy (Void) (Int 4))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 0)
    (Locals
      (Blob 4 4)
      (Int 4))
    (List
      (Block
        (Params (Local 1))
        (Store (Int 4)
          (Addr (Local 0))
          (Copy (Local 1)))
        (Store (Int 4)
          (Addr (Local 0))
          (IntVal 0))
        (Drop
          (Load (Int 4)
            (Addr (Local 0))))
        (Drop
          (Addr (Local 0)))
        (Return))))
  (ProcDef (Type 1)
    (Locals
      (Blob 8 8)
      (Int 8))
    (List
      (Block (Params)
        (Drop
          (Addr (Local 0)))
        (Return))
      (Except
        (Params (Local 1))
        (Store (Int 8)
          (Addr (Local 0))
          (Copy (Local 1)))
        (Return)))))
